# Problem Set 3
## Mahyar Ebrahimitorki

In [2]:
#Cleaning the Data
import pandas as pd 
data1 = pd.read_stata('PS3_data.dta') 
'''
hlabinc = annual labor income of the head
hannhrs = annual hours of the head
hsex = gender of the head (1 = Male, 2 = Female)
hrace = race of the head (1 = white, 2 = Black, 3 = Native American, 4 = Asian/Pacific Islander, 5 = Hispanic, 6,7 = Other)
age = age of the head
hyrsed = years of education of the head
'''
print("Data Statistics:")
data1.describe() 

Data Statistics:


id68           year          intid        hannhrs  \
count  123786.000000  123786.000000  123786.000000  123786.000000   
mean     1494.639475    1984.831273    3271.379429    1679.269897   
std       838.901790       9.836212    2277.056058    1061.704712   
min         1.000000    1967.000000       1.000000       0.000000   
25%       772.000000    1977.000000    1444.000000     832.000000   
50%      1517.000000    1985.000000    2984.000000    1976.000000   
75%      2224.000000    1993.000000    4763.000000    2350.000000   
max      2930.000000    2002.000000   16968.000000    7800.000000   

             wannhrs       hlabinc        wlabinc        nochild  \
count  123786.000000  9.023300e+04   48496.000000  123786.000000   
mean      633.026917  4.211505e+04   22026.289062       0.843771   
std       878.422791  4.670424e+04   21336.107422       1.182829   
min         0.000000  6.353981e-01       1.192780       0.000000   
25%         0.000000  1.979858e+04    8016.247070       0.000000   
50%         0.000000  3.460022e+04   18122.412109       0.000000   
75%      1454.000000  5.267309e+04   30256.060547       2.000000   
max      5840.000000  3.771521e+06  856942.062500      11.000000   

              wrace          hrace  ...  redpregovinc           hsex     wsex  \
count  90603.000000  123656.000000  ...  1.237860e+05  123786.000000  80758.0   
mean       1.098220       1.129731  ...  3.012258e+04       1.233072      2.0   
std        0.356161       0.394627  ...  4.588795e+04       0.422940      0.0   
min        1.000000       1.000000  ... -1.324040e+05       1.000000      2.0   
25%        1.000000       1.000000  ...  7.700000e+03       1.000000      2.0   
50%        1.000000       1.000000  ...  1.900000e+04       1.000000      2.0   
75%        1.000000       1.000000  ...  3.910775e+04       1.000000      2.0   
max        8.000000       8.000000  ...  3.660000e+06       2.000000      2.0   

                 age          wage        hpersno       wpersno  \
count  123786.000000  80758.000000  123786.000000  80758.000000   
mean       45.545547     41.390785      39.620201     55.346169   
std        17.623671     14.786721      69.003265     77.864296   
min        16.000000     13.000000       1.000000      1.000000   
25%        31.000000     29.000000       1.000000      2.000000   
50%        42.000000     39.000000       3.000000      3.000000   
75%        58.000000     51.000000      22.000000    170.000000   
max       102.000000     95.000000     227.000000    231.000000   

              hyrsed        wyrsed            pce  
count  122809.000000  80091.000000  123786.000000  
mean       12.666091     12.720081       0.557690  
std         2.917721      2.422607       0.265198  
min         1.000000      1.000000       0.000000  
25%        12.000000     12.000000       0.362158  
50%        12.000000     12.000000       0.599887  
75%        15.000000     14.000000       0.786908  
max        17.000000     17.000000       0.928007  

[8 rows x 51 columns]

In [3]:
import numpy as np
data2 = data1.dropna(how = 'any', subset = ['hlabinc', 'hannhrs', 'hsex', 'hrace', 'age', 'hyrsed']) #drop missing values
data3 = data2[['id68', 'year', 'intid', 'hlabinc', 'hannhrs', 'hsex', 'hrace', 'age', 'hyrsed']]
data3['ahrs'] = data3['hannhrs'].where(data3['hannhrs']>0)
data3['hrw'] = data3['hlabinc']/data3['ahrs'] 
data3 = data3[(data3.hsex == 1.0) & (data3.age >= 25) & (data3.age <= 60) & (data3.hrw > 7)] # select male heads of HH whose age is between 25 & 60 (included!), and wage > $7/hr
data3['ln_hrw'] = np.log(data3['hrw']) 
race_dummy = pd.get_dummies(data3['hrace']) # Defining race dummies 
data3['constant'] = 1
data = pd.concat([data3, race_dummy], axis = 1) 
data.rename(columns = {1.0: 'White', 2.0: 'Black', 3.0: 'Others'}, inplace = True) 
# Separate years for estimation
data1971 = data[data['year'] == 1971]
data1980 = data[data['year'] == 1980]
data1990 = data[data['year'] == 1990]
data2000 = data[data['year'] == 2000]
data.describe()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


id68          year         intid       hlabinc       hannhrs  \
count  57062.000000  57062.000000  57062.000000  5.706200e+04  57062.000000   
mean    1507.470558   1986.575672   3480.375311  5.282805e+04   2228.557617   
std      828.361439      8.712311   2253.229068  5.236477e+04    620.054077   
min        1.000000   1971.000000      1.000000  1.666980e+01      2.000000   
25%      782.000000   1979.000000   1690.000000  3.037345e+04   1952.000000   
50%     1542.000000   1987.000000   3296.000000  4.382858e+04   2160.000000   
75%     2225.000000   1994.000000   5002.000000  6.138495e+04   2519.000000   
max     2930.000000   2002.000000  16968.000000  3.771521e+06   5840.000000   

          hsex         hrace           age        hyrsed          ahrs  \
count  57062.0  57062.000000  57062.000000  57062.000000  57062.000000   
mean       1.0      1.101416     39.243629     13.529967   2228.557617   
std        0.0      0.369015      9.578858      2.450013    620.054077   
min        1.0      1.000000     25.000000      1.000000      2.000000   
25%        1.0      1.000000     31.000000     12.000000   1952.000000   
50%        1.0      1.000000     38.000000     13.000000   2160.000000   
75%        1.0      1.000000     47.000000     16.000000   2519.000000   
max        1.0      3.000000     60.000000     17.000000   5840.000000   

                hrw        ln_hrw  constant         White         Black  \
count  57062.000000  57062.000000   57062.0  57062.000000  57062.000000   
mean      24.320921      3.010804       1.0      0.921103      0.056377   
std       25.209909      0.544096       0.0      0.269580      0.230651   
min        7.000252      1.945946       1.0      0.000000      0.000000   
25%       13.950662      2.635527       1.0      1.000000      0.000000   
50%       19.914677      2.991457       1.0      1.000000      0.000000   
75%       27.790929      3.324710       1.0      1.000000      0.000000   
max     1717.330322      7.448526       1.0      1.000000      1.000000   

             Others  
count  57062.000000  
mean       0.022519  
std        0.148367  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [4]:
data.dtypes

id68          int16
year          int16
intid         int16
hlabinc     float32
hannhrs     float32
hsex        float32
hrace       float64
age         float32
hyrsed      float32
ahrs        float32
hrw         float32
ln_hrw      float32
constant      int64
White         uint8
Black         uint8
Others        uint8
dtype: object

In [5]:
#MLE
'''
ln(w_it) = alpha + beta_1 * Educ_it + beta_2 * Age_it + beta_3 * Black_it + beta_4 * Hispanic_it + beta_5 * OtherRace_it + epsilon_it
w_it = wage of individual i in survey year t
Educ_it = education in years
Age_it = age in years
Black_it, Hispanic_it, OtherRace_it = dummy variables for race = Black, Hispanic, Not (belongs to {White, Black, Hispanic})
'''
import statsmodels.api as statmod 
import scipy.stats as stats

def Log_Lik(params, t): #define negative maximum likelihood
    # Coeff.s
    beta0, beta1, beta2, beta3, beta4, sigma = params
    beta = np.array([beta0, beta1, beta2, beta3, beta4])
    n = len(t)
    # Independent variables matrix (No Hispanic)
    x0 = np.array(t['constant']).astype('float')
    x1 = np.array(t['hyrsed']).astype('float')
    x2 = np.array(t['age']).astype('float')
    x3 = np.array(t['Black']).astype('float')
    x4 = np.array(t['Others']).astype('float')
    X = np.column_stack((x0, x1, x2, x3, x4))
    # Dependent variable matrix
    y = np.array(t['ln_hrw']).astype('float')
    y_bar = np.dot(X, beta)
    ll = (-(n/2)*np.log(2*np.pi) - (n/2)*np.log(sigma**2) - (1/(2*sigma**2))*((y-y_bar).T @ (y - y_bar)))
    return (-ll)


In [24]:
from scipy.optimize import minimize 
import scipy.optimize as opt

# MLE; 'Nelder-Mead'
b0 = 1.5
b1 = 0.1
b2 = 0.01
b3 = 0.2
b4 = 0.01
SD = 0.5
beta = [b0, b1, b2, b3, b4]
params = [b0, b1, b2, b3, b4, SD]
bounds = ((1e-10, None), (None,None), (None,None), (None,None), (None,None), (None,None))

print("Coefficients: ", "[b0  b1  b2  b3  b4  SD]")
print("Whole Data: ", opt.minimize(Log_Lik, params, args=(data), method='Nelder-Mead', bounds=bounds).x)
print("year = 1971: ", opt.minimize(Log_Lik, params, args=(data1971), method='Nelder-Mead', bounds=bounds).x)
print("year = 1980: ", opt.minimize(Log_Lik, params, args=(data1980), method='Nelder-Mead', bounds=bounds).x)
print("year = 1990: ", opt.minimize(Log_Lik, params, args=(data1990), method='Nelder-Mead', bounds=bounds).x)
print("year = 2000: ", opt.minimize(Log_Lik, params, args=(data2000), method='Nelder-Mead', bounds=bounds).x)

Coefficients:  [b0  b1  b2  b3  b4  SD]
Whole Data:  [ 1.39248744  0.07813422  0.01452121 -0.16203288  0.01954363  0.48968626]
year = 1971:  [ 1.55093966  0.06687836  0.01439224 -0.16391561  0.03071908  0.41009271]
year = 1980:  [ 1.61284465  0.06756298  0.01269706 -0.10239978  0.02204488  0.44923944]
year = 1990:  [ 1.11858534  0.09755627  0.01346643 -0.17202102 -0.05972682  0.48360497]
year = 2000:  [ 1.15653207  0.10915031  0.01105583 -0.24282111  0.01545989  0.53974426]


In [25]:
# MLE; 'L-BFGS-B'
b0 = 1.5
b1 = 0.1
b2 = 0.01
b3 = 0.2
b4 = 0.01
SD = 0.5
beta = [b0, b1, b2, b3, b4]
params = [b0, b1, b2, b3, b4, SD]
bounds = ((1e-10, None), (None,None), (None,None), (None,None), (None,None), (None,None))

print("Coefficients: ", "[b0  b1  b2  b3  b4  SD]")
print("Whole dataset: ", opt.minimize(Log_Lik, params, args=(data), method='L-BFGS-B', bounds=bounds).x)
print("year = 1971: ", opt.minimize(Log_Lik, params, args=(data1971), method='L-BFGS-B', bounds=bounds).x)
print("year = 1980: ", opt.minimize(Log_Lik, params, args=(data1980), method='L-BFGS-B', bounds=bounds).x)
print("year = 1990: ", opt.minimize(Log_Lik, params, args=(data1990), method='L-BFGS-B', bounds=bounds).x)
print("year = 2000: ", opt.minimize(Log_Lik, params, args=(data2000), method='L-BFGS-B', bounds=bounds).x)

Coefficients:  [b0  b1  b2  b3  b4  SD]
Whole dataset:  [ 1.3926374   0.07813449  0.01451934 -0.1622223   0.01596499  0.48968195]
year = 1971:  [ 1.5509634   0.06687915  0.0143912  -0.16387764  0.03068677  0.4100968 ]
year = 1980:  [ 1.6131477   0.06755473  0.01269749 -0.10274478  0.01344393  0.44924268]
year = 1990:  [ 1.11784801  0.09759148  0.01347082 -0.1718518  -0.05612097  0.48360277]
year = 2000:  [ 1.16169768  0.10915415  0.01099341 -0.2460413  -0.06072644  0.53955648]


In [26]:
# MLE; 'SLSQP'
b0 = 1.5
b1 = 0.1
b2 = 0.01
b3 = 0.2
b4 = 0.01
SD = 0.5
beta = [b0, b1, b2, b3, b4]
params = [b0, b1, b2, b3, b4, SD]
bounds = ((1e-10, None), (None,None), (None,None), (None,None), (None,None), (None,None))

print("Coefficients: ", "[b0  b1  b2  b3  b4  SD]")
print("Total dataset: ", opt.minimize(Log_Lik, params, args=(data), method='SLSQP', bounds=bounds).x)
print("year = 1971: ", opt.minimize(Log_Lik, params, args=(data1971), method='SLSQP', bounds=bounds).x)
print("year = 1980: ", opt.minimize(Log_Lik, params, args=(data1980), method='SLSQP', bounds=bounds).x)
print("year = 1990: ", opt.minimize(Log_Lik, params, args=(data1990), method='SLSQP', bounds=bounds).x)
print("year = 2000: ", opt.minimize(Log_Lik, params, args=(data2000), method='SLSQP', bounds=bounds).x)


Coefficients:  [b0  b1  b2  b3  b4  SD]
Total dataset:  [ 1.3930103   0.07811907  0.0145183  -0.16240227  0.01105441  0.48968219]
year = 1971:  [ 1.55096656  0.0668779   0.01439154 -0.16387704  0.03068453  0.41009888]
year = 1980:  [ 1.61309174  0.06755659  0.0126982  -0.10273663  0.01350612  0.44924221]
year = 1990:  [ 1.11859812  0.09755771  0.01346542 -0.17202113 -0.05971542  0.48359832]
year = 2000:  [ 1.16169979  0.10915436  0.0109933  -0.2460477  -0.06073881  0.53955706]


In [28]:
# OLS 
print('OLS; Whole data',  statmod.OLS(endog=data['ln_hrw'], exog=data[['constant', 'hyrsed', 'age', 'Black', 'Others']]).fit().summary())
print('OLS; year = 1971', statmod.OLS(endog=data1971['ln_hrw'], exog=data1971[['constant', 'hyrsed', 'age', 'Black', 'Others']]).fit().summary())
print('OLS; year = 1980', statmod.OLS(endog=data1980['ln_hrw'], exog=data1980[['constant', 'hyrsed', 'age', 'Black', 'Others']]).fit().summary())
print('OLS; year = 1990', statmod.OLS(endog=data1990['ln_hrw'], exog=data1990[['constant', 'hyrsed', 'age', 'Black', 'Others']]).fit().summary())
print('OLS; year = 2000', statmod.OLS(endog=data2000['ln_hrw'], exog=data2000[['constant', 'hyrsed', 'age', 'Black', 'Others']]).fit().summary())

OLS; Whole data                             OLS Regression Results                            
Dep. Variable:                 ln_hrw   R-squared:                       0.190
Model:                            OLS   Adj. R-squared:                  0.190
Method:                 Least Squares   F-statistic:                     3346.
Date:                Sat, 28 Sep 2019   Prob (F-statistic):               0.00
Time:                        22:33:15   Log-Likelihood:                -40225.
No. Observations:               57062   AIC:                         8.046e+04
Df Residuals:                   57057   BIC:                         8.051e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
constant       1.3930      0.015    

# Interpretation 


The coefficients of education in all the above estimations are positive and progressive from 1971 to 2000 based. As the desired model is linear, the results for OLS estimation and MLE match and are the same in magnitude and sign. That is, one additional year of education for male heads increases their wages by 7.8% in all estimations for the whole sample. The results for the other samples of the data (4 years of 1971, 1980, 1990, and 2000) are also very close to the OLS estimation in these samples. 